### ISSUE TITLE
Add dropout to NeuralNetwork class

### ISSUE THREAD

**Issue by hideyukiinada** [hideyukiinada/ml#14]
Add dropout to NeuralNetwork class

**Comment by hideyukiinada** [comment]
Current thought is:
1. Randomly pick a subset of weights for the layer where dropout is specified.
2. Save those weights
3. Set the value of those weights to 0
4. Front prop (those weights should not be added while calculating z thus making them invisible)
5. Back prop (During update_weights(), grad is subtracted from those weights resulting in bogus values, but that's OK as they will be reset before the next front prop).
6. Restore the original value of weights

**Comment by hideyukiinada** [comment]
I also need to figure out how dropout works for a conv layer.

**Comment by hideyukiinada** [comment]
Implemented for a hidden layer.  Final layer and input layer are not supported.
See:
https://github.com/hideyukiinada/ml/blob/master/examples/neural_network_mnist_sgd_dropout_example
I used this paper as the reference:
https://www.cs.toronto.edu/~hinton/absps/JMLRdropout.pdf

The logic is:
1. During initialization, for each layer, allocate a vector that matches the number of activations for the layer.
2. During front prop,
  Initialize the vector with all 1s.
  Randomly pick a subset of the elements of the vector for the layer where dropout is specified.  Use the (1 - dropout) as the ratio to determine the size of the subset.
3. During back prop, multiply dL/da_prev with the vector.
